lua/entities/gmod_wire_expression2/core/custom/nav.lua

```diff
@@ -0,0 +1,335 @@
+local MAP_HAS_NAV = navmesh.IsLoaded()
+hook.Add("PlayerInitialSpawn","__e2_MAP_HAS_NAV",function()
+	MAP_HAS_NAV = navmesh.IsLoaded()
+	hook.Remove("PlayerInitialSpawn","__e2_MAP_HAS_NAV")
+end )
+
+local nullMesh = navmesh.GetNavAreaByID(0)
+local nullVec = Vector()
+local nullTable = {}
+
+local Clamp = math.Clamp
+local floor = math.floor
+
+--------------------
+
+registerType("navmesh", "xnv", nullMesh,
+	nil,
+	nil,
+	function(retval)
+		if retval == nil then return end
+		if IsValid(retval) then return end
+		local _type = type(retval)
+		if _type~="CNavArea" then error("Return value is neither nil nor a CNavArea, but a "..type(retval).."!",0) end
+	end,
+	function(v)
+		return type(v)~="CNavArea"
+	end
+)
+
+registerOperator("ass", "xnv", "xnv", function(self, args)
+	local op1, op2, scope = args[2], args[3], args[4]
+	local      rv2 = op2[1](self, op2)
+	self.Scopes[scope][op1] = rv2
+	self.Scopes[scope].vclk[op1] = true
+	return rv2
+end )
+
+e2function number operator_is(navmesh nav)
+	if IsValid(nav) then return 1 else return 0 end
+end
+
+e2function number operator==(navmesh lhs, navmesh rhs)
+	if lhs == rhs then return 1 else return 0 end
+end
+
+e2function number operator!=(navmesh lhs, navmesh rhs)
+	if lhs ~= rhs then return 1 else return 0 end
+end
+
+--------------------
+
+E2Lib.RegisterExtension("nav", false, "Read navmesh data from the map if it has one","Experimental, may be unreasonably costly.")
+
+--------------------
+
+local function NavBounds(nav)
+	if not IsValid(nav) then return nullTable end
+	local t = nullTable
+	local c = nav:GetCenter()
+	local mx, my, mz, MX, MY, MZ = c.x, c.y, c.z, c.x, c.y, c.z
+	for i=1,4 do
+		t[i] = nav:GetCorner(i-1)
+		mx = math.min(t[i].x, mx)
+		my = math.min(t[i].y, my)
+		mz = math.min(t[i].z, mz)
+		MX = math.max(t[i].x, MX)
+		MY = math.max(t[i].y, MY)
+		MZ = math.max(t[i].z, MZ)
+	end
+	return { Vector(mx, my, mz), Vector(MX, MY, MZ) }
+end
+
+-------------------- TODO: refactor/rewrite Astar to remove more garrycode such as continue
+
+function Astar(start, goal)

```

Comment:
I'm also still concerned about infinite loop here. Are we absolutely sure this has no infinite loop?

Reply:
just so we're clear, we're looking at the astar right? not the navbounds?
honestly would have never dropped that in if it weren't conveniently available on the wiki for public use. i barely knew how it worked in the first place and despite understanding what the closed/open lists are now i still kind of don't, just surprised it actually would cause a problem until I looked up others seemingly have that situation outside of garry's mod game dev.

at this point, its either drop it and let e2 handle pathfinding or to blindly trust whoever added that to the wiki. it's been sane in the couple tests i've used it in so far.

Reply:
i'll take a look at what table does for dynamic op costs yeah. think it was tables anyway.

Reply:
all of the functions used currently are part of gmod's standard navmesh functions so if they break, blame garry.

just like i did with Nav:GetExtent()

Reply:
to answer your question directly: no

Reply:
working on the iterator